### PR TITLE
[config] Add nullness annotations to config validation

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added nullness annotations to config validation
- Use constructor injection

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java [hunk 1]
```diff
@@ -49,20 +47,28 @@
...
+    private final Logger logger = LoggerFactory.getLogger(ConfigDescriptionValidatorImpl.class);
 
-    private static final List<ConfigDescriptionParameterValidator> VALIDATORS = Collections
-            .unmodifiableList(Arrays.asList(ConfigDescriptionParameterValidatorFactory.createRequiredValidator(),
-                    ConfigDescriptionParameterValidatorFactory.createTypeValidator(),
-                    ConfigDescriptionParameterValidatorFactory.createMinMaxValidator(),
-                    ConfigDescriptionParameterValidatorFactory.createPatternValidator()));
+    private final ConfigDescriptionRegistry configDescriptionRegistry;
+    private final TranslationProvider translationProvider;
+    private @NonNullByDefault({}) BundleContext bundleContext;
```
Comment (J-N-K) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:64]: Why don't you inject that with the constructor, too?

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java [hunk 2]
```diff
@@ -49,20 +47,28 @@
...
 
-    private BundleContext bundleContext;
+    @Activate
+    public ConfigDescriptionValidatorImpl(final @Reference ConfigDescriptionRegistry configDescriptionRegistry,
+            final @Reference TranslationProvider translationProvider) {
```
Comment (J-N-K) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:68]: This makes the `TranslationProvider` and the `ConfigDescriptionRegistry` mandatory, they were optional before. That change makes sense to me, but did you check if this does not create circular dependencies?

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java (hunk 1 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:64]: Ups, ... Will add it.

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java (hunk 2 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:68]: No, but afair the `TranslationProvider` does not reference any other services. The `ConfigDescriptionRegistry` only references optional provider.

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java [hunk 3]
```diff
@@ -86,7 +83,7 @@ protected void deactivate() {
      * @throws NullPointerException if given config description URI or configuration parameters are null
      */
     @Override
-    @SuppressWarnings("unchecked")
+    @SuppressWarnings({ "unchecked", "null" })
     public void validate(Map<String, Object> configurationParameters, URI configDescriptionURI) {
         Objects.requireNonNull(configurationParameters, "Configuration parameters must not be null");
```
Comment (J-N-K) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:88]: If the class is annotated `@NonNullByDefault` both parameters are `@NonNull`. Isn't this check unnecessary?

**Review by J-N-K — COMMENTED** [review]
In my opinion we can also remove `ConfigDescriptionParameterValidatorFactory`. I don't think there is any benefit in calling the factory compared to instantiate directly. If the idea was to save resources, it would be better to create a single static instance of each validator.

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java (hunk 3 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:88]: Yes, and there are plenty more of them which can be omitted in the related validation classes. But let us focus on annotations and c-tor injection here and move other improvements into a follow-up PR.

Comment (J-N-K) [reply bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:88]: Fine with me. I don't understand why the build failed.

**Comment by cweitkamp** [comment]
@openhab/core-maintainers May I ask you to have a look into this PR. Thanks.

File: bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java [hunk 4]
```diff
@@ -49,29 +46,29 @@
...
+    private final Logger logger = LoggerFactory.getLogger(ConfigDescriptionValidatorImpl.class);
 
-    private BundleContext bundleContext;
+    private final ConfigDescriptionRegistry configDescriptionRegistry;
+    private final TranslationProvider translationProvider;
+    private final BundleContext bundleContext;
 
     @Activate
-    protected void activate(BundleContext bundleContext) {
+    public ConfigDescriptionValidatorImpl(final @Reference BundleContext bundleContext,
```
Comment (wborn) [inline bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java:66]: You can probably fix the build by removing `@Reference` here:

```suggestion
    public ConfigDescriptionValidatorImpl(final BundleContext bundleContext,
```

The `BundleContext` is not something you can get a reference on like another component. But I think the annotations processor is smart enough to inject the `BundleContext` when it sees such a parameter. We also use it like this with the `StartLevelService`.

Comment (cweitkamp) [reply bundles/org.openhab.core.config.core/src/main/java/org/openhab/core/config/core/internal/validation/ConfigDescriptionValidatorImpl.java]: Thanks for the hint. I applied your suggestion.

**Pull request merged**